### ISSUE TITLE
DesMume AVX2 build screen pixel shift.

### ISSUE THREAD

**Issue by ppsspp-gamer** [TASVideos/desmume#8]
I compiled a DesMume AVX2 build with vs2015 by changing "desmume.props" SSE_Level to 40 and AVX_Level to 20. The program ran fine except screen pixel shift as picture below. 

![avx2](https://cloud.githubusercontent.com/assets/8726272/21082019/65be5594-c00d-11e6-8e81-c4c4184e39ae.jpg)

**Comment by zeromus** [comment]
relatedly, since we now have the libretro-common cpu features checker, we could enable some of the CPU extensions at runtime regardless of what the build is compiled to use, couldnt we?

**Comment by rogerman** [comment]
No, we can't dynamically use different CPU extensions at runtime because there are a ton of C++ objects that are globally instantiated on the stack at startup. Since the C++ std library only compiles with the highest level CPU extensions available, this will cause crashes on CPUs without those extensions.

We would have to reengineer DeSmuME to eliminate all globally instantiated C++ objects. Not an easy feat due to the sheer number of them that exist.

**Comment by zeromus** [comment]
I don't think you meant stack. 

you can reconstruct globals at runtime, if you're careful and rewire them all correctly. Heck, you can do that on the stack too. But I can see a lot of that code for instance in GPU is conditionally compiled.

But I thought for the stuff which is more modularized (the colorspace handler) it would work. Just the way it is, only declare two and select at runtime (they already use an interface base class ColorspaceHandler) which one (and you can drop a ColorspaceHandler* pointer to use as the interface after the CPU is selected)

Of course this is all easy for me to say, I won't be the one doing it.

And I think I wasn't clear about what I proposed.

We could make it so the desmume would only be compiled with SSE support (and the c++ std library and compiler-generated instructions) would wall be SSE at the most. Then, in the cases where we do have the modularization already done, we can select it dynamically. I really don't see why we can't--I've had people telling me for years only dummies can't select CPU instruction sets at runtime (theyre the dummies, since there's other considerations)--but still, they way THEY do things IS by selecting modular interfaces at runtime (and apparently forgoing lots of other more compiler-automatic optimizations since they dont even seem to know about it)

**Comment by rogerman** [comment]
Oh, whoops, I was thinking about something else stack-related while I was typing that. My bad.

But still, I don't want to commit to any runtime CPU extension selection stuff just yet. I'm trying to get other stuff accomplished at the moment.

As for this issue, this does seem to be a data alignment bug, which probably snuck in since I didn't fully test the AVX2 code. (I don't have easy access to any AVX2-capable CPUs.) I have a feeling that that the bug is probably in _ConvertColorBaseTo5551_AVX2(), but I'll need to find some time to get this tested properly.

**Comment by ductator** [comment]
zeromus PR fixed compile error with AVX2, but graphics error still exist.
`1>GPU.obj : error LNK2001: Nicht aufgelöstes externes Symbol ""void __cdecl ColorspaceConvert555To6665Opaque_SSE2<0>(union __m128i const &,union __m128i &,union __m128i &)" (??$ColorspaceConvert555To6665Opaque_SSE2@$0A@@@YAXAEBT__m128i@@AEAT0@1@Z)".
1>GPU.obj : error LNK2001: Nicht aufgelöstes externes Symbol ""void __cdecl ColorspaceConvert555To8888Opaque_SSE2<0>(union __m128i const &,union __m128i &,union __m128i &)" (??$ColorspaceConvert555To8888Opaque_SSE2@$0A@@@YAXAEBT__m128i@@AEAT0@1@Z)".
1>OGLRender.obj : error LNK2001: Nicht aufgelöstes externes Symbol ""union __m128i __cdecl ColorspaceConvert8888To6665_SSE2<1>(union __m128i const &)" (??$ColorspaceConvert8888To6665_SSE2@$00@@YA?AT__m128i@@AEBT0@@Z)".
1>OGLRender.obj : error LNK2001: Nicht aufgelöstes externes Symbol ""union __m128i __cdecl ColorspaceConvert8888To5551_SSE2<1>(union __m128i const &,union __m128i const &)" (??$ColorspaceConvert8888To5551_SSE2@$00@@YA?AT__m128i@@AEBT0@0@Z)".
1>texcache.obj : error LNK2001: Nicht aufgelöstes externes Symbol ""void __cdecl ColorspaceConvert555To8888_SSE2<0>(union __m128i const &,union __m128i const &,union __m128i const &,union __m128i &,union __m128i &)" (??$ColorspaceConvert555To8888_SSE2@$0A@@@YAXAEBT__m128i@@00AEAT0@1@Z)".
1>texcache.obj : error LNK2001: Nicht aufgelöstes externes Symbol ""void __cdecl ColorspaceConvert555To6665_SSE2<0>(union __m128i const &,union __m128i const &,union __m128i const &,union __m128i &,union __m128i &)" (??$ColorspaceConvert555To6665_SSE2@$0A@@@YAXAEBT__m128i@@00AEAT0@1@Z)".
`
This is the error, that failed the compiling before cimmit 5cfb407. AVX2 activated via DeSmuME-user.props.
AVX2 compiles used to work flawlessly but were broken on the old sourceforge repo at some point in time. Unfortunately I can't pinpoint the exact commit, which broke AVX2 support.

**Comment by rogerman** [comment]
Okay, I can confirm the AVX2 bug, and I've also found where it is: It's the vpunpcklwd and vpunpckhwd instructions that I used in ColorspaceConvert555To8888_AVX2() and ColorspaceConvert555To6665_AVX2().

Apparently, these instructions don't work the way I originally thought they did. I thought they worked on the entire 256-bit register at once, but they actually work in only 128-bit chunks within the 256-bit register. This is why the pixels are getting swapped in the positioning.

Unfortunately, I can't simply move the pixels back into proper positioning because neither the vpshufd or vpshufb instructions allow you to move bytes across 128-bit lanes. This means that I will need to rewrite the functions without using vpunpcklwd or vpunpckhwd. 

The vpmovzxwd instruction seems to look like the promising candidate for the rework.
